v3/lgr_hyper_ep.hpp

```diff
@@ -0,0 +1,648 @@
+        conv = 1;
+        break;
+      } else if (iter > 24 && f <= tol1 * 1000.0) {
+        // Weaker convergence
+        conv = 2;
+        break;
+      }
+#ifdef LGR_HYPER_EP_VERBOSE_DEBUG
+      std::cout << "Iteration: " << iter + 1 << "\n"
+                << "\tROOTJ20: " << hpc::norm(S0) << "\n"
+                << "\tROOTJ2: " << hpc::norm(S) << "\n"
+                << "\tep: " << ep << "\n"
+                << "\tepdot: " << epdot << "\n"
+                << "\tgamma: " << gamma << "\n"
+                << "\tg: " << g << "\n"
+                << "\tdg: " << dg << "\n\n\n";
+#endif
+    }
+    // Update the stress tensor
+    T = Te - twomu * gamma * N;
+    if (!conv) {
+      return ErrorCode::RADIAL_RETURN_FAILURE;
+    } else if (conv == 2) {
+      // print warning about weaker convergence
+    }
+    // Update damage
+    dp = scalar_damage(props, T, dp, temp, ep, epdot, dtime);
+  }
+
+  if (flag != StateFlag::ELASTIC) {
+    // determine elastic deformation
+    auto const jac = determinant(F);
+    auto const Bbe = find_bbe(T, mu);
+    auto const j13 = std::cbrt(jac);
+    auto const j23 = j13 * j13;
+    auto const Be = Bbe * j23;
+    auto const Ve = sqrt_spd(Be);
+    Fp = invert(Ve) * F;
+    if (flag == StateFlag::REMAPPED) {
+      // Correct pressure term
+      auto p = trace(T);
+      auto const D1 = 6.0 * (1.0 - 2.0 * Nu) / E;
+      p = (2.0 * jac / D1 * (jac - 1.0)) - (p / 3.0);
+      for (int i = 0; i < 3; ++i) T(i, i) = p;
+    }
+  }
+  return ErrorCode::SUCCESS;
+}
+
+HPC_NOINLINE inline
+hpc::symmetric_stress<double>
+linear_elastic_stress(Properties const props, hpc::deformation_gradient<double> const Fe)
+{
+  auto const E = props.E;
+  auto const nu = props.Nu;
+  auto const K = E / (3.0 * (1.0 - 2.0 * nu));
+  auto const G = E / 2.0 / (1.0 + nu);
+  auto const grad_u = Fe - hpc::deformation_gradient<double>::identity();
+  auto const strain = (1.0 / 2.0) * hpc::symmetric_deformation<double>(grad_u + transpose(grad_u));
+  auto const isotropic_strain = iso(strain);
+  auto const deviatoric_strain = deviator(strain);
+  return (3.0 * K) * isotropic_strain + (2.0 * G) * deviatoric_strain;
+}
+
+/*
+ * Update the stress using Neo-Hookean hyperelasticity
+ *
+ */
+HPC_NOINLINE inline
+hpc::symmetric_stress<double>
+hyper_elastic_stress(Properties const props, hpc::deformation_gradient<double> const Fe, double const jac)
+{
+  auto const E = props.E;
+  auto const Nu = props.Nu;
+  // Jacobian and distortion tensor
+  auto const scale = 1.0 / std::cbrt(jac);
+  auto const Fb = scale * Fe;
+  // Elastic moduli
+  auto const C10 = E / (4.0 * (1.0 + Nu));
+  auto const D1 = 6.0 * (1.0 - 2.0 * Nu) / E;
+  auto const EG = 2.0 * C10 / jac;
+  // Deviatoric left Cauchy-Green deformation tensor
+  auto Bb = Fb * transpose(Fb);
+  // Deviatoric Cauchy stress
+  auto const TRBb = trace(Bb) / 3.0;
+  for (int i = 0; i < 3; ++i) Bb(i, i) -= TRBb;
+  auto T = hpc::symmetric_stress<double>(EG * Bb);
```

Comment:
@ibaned can we attach units to constants so that the above expression is simply `auto T = EG * Bb`, where `EG` has units of stress and `Bb` is unitless (and `T` ends up with units of stress)?  That would shut be nice!  As is, clang 7.0.1 complains unless I explicitly cast `T`.

Reply:
I think I got that to happen for expressions where `EG` is a scalar, but its not general enough to do it with `EG` as tensor maybe... I'll think about it.

Reply:
In this case `EG` is a scalar with stress units and `Bb` is a tensor with deformation (non dimensional) units.  When multiplied, the units of `EG` should broadcast to each component of `Bb`.  Or something like that!